### PR TITLE
Logical delete of schemas

### PR DESCRIPTION
**Pull request by alxdarksage**

### CODE REVIEW

File: test/org/sagebionetworks/bridge/models/upload/UploadSchemaTest.java [hunk 1]
```diff
@@ -282,6 +283,7 @@ public void testSerialization() throws Exception {
         assertEquals("survey-guid", jsonNode.get("surveyGuid").textValue());
         assertEquals("UploadSchema", jsonNode.get("type").textValue());
         assertEquals(6,  jsonNode.get("version").intValue());
+        assertTrue(jsonNode.get("deleted").booleanValue());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/upload/UploadSchemaTest.java]: Should we also test that the POJO has deleted=true?

File: test/org/sagebionetworks/bridge/play/controllers/UploadSchemaControllerTest.java [hunk 2]
```diff
@@ -76,18 +79,53 @@ public void createSchema() throws Exception {
...
+    @Test
+    public void deleteAllRevisionsOfUploadSchemaPermanently() throws Exception {
+        UploadSchemaService mockSvc = mock(UploadSchemaService.class);
 
+        // setup, execute, and validate
+        UploadSchemaController controller = setupControllerWithService(mockSvc, Roles.DEVELOPER, Roles.ADMIN);
+        
+        Result result = controller.deleteAllRevisionsOfUploadSchema("delete-schema", "true");
+        TestUtils.assertResult(result, 200, "Schemas have been deleted.");
+        verify(mockSvc).deleteUploadSchemaByIdPermanently(TestConstants.TEST_STUDY, "delete-schema");
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/UploadSchemaControllerTest.java:102]: What about when physical=true, but the caller is not an admin? (Here and deleteSchemaRevision.)

File: app/org/sagebionetworks/bridge/services/UploadSchemaService.java [hunk 3]
```diff
@@ -547,42 +559,42 @@ public void deleteUploadSchemaById(StudyIdentifier studyId, String schemaId) {
...
-
-        List<SharedModuleMetadata> sharedModuleMetadataList = sharedModuleMetadataService.queryAllMetadata(false, false,
-                "schemaId=:schemaId AND schemaRevision=:schemaRevision", parameters, null);
-
-        if (sharedModuleMetadataList.size() != 0) {
-            throw new BadRequestException("Cannot delete specified Upload Schema because a shared module still refers to it.");
+        UploadSchema schema = getRevisionForDeletion(studyId, schemaId, rev);
+        if (schema == null || schema.isDeleted()) {
+            throw new EntityNotFoundException(UploadSchema.class);
         }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/UploadSchemaService.java:575]: Unit test for deleting a deleted schema?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
Overall looks good, a few notes.

File: test/org/sagebionetworks/bridge/models/upload/UploadSchemaTest.java (hunk 1 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/upload/UploadSchemaTest.java]: Yes, added

File: test/org/sagebionetworks/bridge/play/controllers/UploadSchemaControllerTest.java (hunk 2 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/UploadSchemaControllerTest.java:102]: Added for both forms of delete.

File: app/org/sagebionetworks/bridge/services/UploadSchemaService.java (hunk 3 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/UploadSchemaService.java:575]: Added test that logically deleting logically deleted schema throws ENFE, added test that physically deleting a logically deleted schema succeeds.

**Review by DwayneJengSage — APPROVED** [review]

**Pull request merged**